File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        
        int n = nums.length;
        Integer[][] dp = new Integer[n+1][n+1];
        return solve(0,-1,nums,dp);
    }
    
    
    int solve(int idx,int prev,int[]nums,Integer[][] dp){
        
        if(idx>=nums.length)return 0;
        if(prev != -1 && dp[idx][prev] != null)return dp[idx][prev];
        int np = solve(idx+1,prev,nums,dp);
        
        int pick = 0;
        
        if(prev == -1 || nums[idx] > nums[prev]){
            pick  = 1 + solve(idx+1,idx,nums,dp); 
        }
        if(prev==-1) return Math.max(pick,np);
        return dp[idx][prev] = Math.max(pick,np);
        
    }
    //     int n = nums.length;
    //     int[] dp = new int[n];
    //     int ans  = 0;
    //     for(int i = 0;i<n;i++){
    //         int max = 0;
    //         for(int j = 0;j<i;j++){
    //             if(nums[i] > nums[j]){
    //                 max = Math.max(dp[j],max);
    //             }   
    //         }  
    //         dp[i] = max+1;
    //         ans = Math.max(ans,dp[i]);
    //     }
    //     return ans;
    // }
    
    

     // public int lengthOfLIS(int[] nums) {
     //     int n = nums.length;
     //     ArrayList<Integer> al = new ArrayList<>();
     //     // al.add(nums[0]);
     //    int ans= 0;
     //     for(int i = 0;i<n;i++){
     //         if( al.isEmpty() || nums[i] > al.get(al.size()-1) ){
     //             al.add(nums[i]);
     //             ans++;
     //         }else{
     //             int pos =  Collections.binarySearch(al,nums[i]);
     //             if(pos<0){
     //                 pos = Math.abs(pos+1);
     //                    al.set(pos,nums[i]);
     //             }
     //         }  
     //     }
     //     return ans;
     // }
}

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.cpp
class Solution {
public:
    int f(int idx, int prev_idx, int n, vector<vector<int>>&dp, vector<int>&nums){
        if(idx==n){
            return 0;
        }
        if(dp[idx][prev_idx+1]!=-1){
            return dp[idx][prev_idx+1];
        }
        int not_take=0+f(idx+1,prev_idx,n,dp,nums);
        int take=INT_MIN;
        if(prev_idx==-1 || nums[idx]>nums[prev_idx]){
            take=1+f(idx+1,idx,n,dp,nums);
        }
        return dp[idx][prev_idx+1]=max(take,not_take);
    }
    int lengthOfLIS(vector<int>& nums) {
        int n=nums.size();
        vector<vector<int>>dp(n+1,vector<int>(n+1,-1));
        return f(0,-1,n,dp,nums);
    }
};